arrow-libs/core/arrow-core-data/src/main/kotlin/arrow/core/NonEmptyList.kt

```diff
@@ -550,8 +429,188 @@ object NonEmptyListSemigroup : Semigroup<NonEmptyList<Any?>> {
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  map: (A, B, C, D) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, _, _, _, _, _, _ ->
+    map(aa, bb, cc, dd)
+  }
+
+inline fun <A, B, C, D, E, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  map: (A, B, C, D, E) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, _, _, _, _, _ ->
+    map(aa, bb, cc, dd, ee)
+  }
+
+inline fun <A, B, C, D, E, F, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  map: (A, B, C, D, E, F) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, _, _, _, _ ->
+    map(aa, bb, cc, dd, ee, ff)
+  }
+
+inline fun <A, B, C, D, E, F, G, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  map: (A, B, C, D, E, F, G) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g,
+    NonEmptyList.unit,
+    NonEmptyList.unit,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, _, _, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  map: (A, B, C, D, E, F, G, H) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g, h,
+    NonEmptyList.unit, NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, hh, _, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg, hh)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, I, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  i: NonEmptyList<I>,
+  map: (A, B, C, D, E, F, G, H, I) -> Z
+): NonEmptyList<Z> =
+  zip(
+    b, c, d, e, f, g, h, i,
+    NonEmptyList.unit
+  ) { aa, bb, cc, dd, ee, ff, gg, hh, ii, _ ->
+    map(aa, bb, cc, dd, ee, ff, gg, hh, ii)
+  }
+
+inline fun <A, B, C, D, E, F, G, H, I, J, Z> NonEmptyList<A>.zip(
+  b: NonEmptyList<B>,
+  c: NonEmptyList<C>,
+  d: NonEmptyList<D>,
+  e: NonEmptyList<E>,
+  f: NonEmptyList<F>,
+  g: NonEmptyList<G>,
+  h: NonEmptyList<H>,
+  i: NonEmptyList<I>,
+  j: NonEmptyList<J>,
+  map: (A, B, C, D, E, F, G, H, I, J) -> Z
+): NonEmptyList<Z> {
+  val buffer = ArrayList<Z>()
+  for (aa in tail) {
```

Comment:
this should iterate over this and the below over the nonemptylist entirely, not just the tail

```suggestion
  for (aa in this) {
```

Although now that I'm thinking of it, is there a reason why we need mapN here at all? We could simply rely on the Iterable version as it would do the same, no?

Reply:
Yes, we actually just use `NonEmptyList.unsafeFromList(all.zip(other.all, f))`